Mathlib/Topology/Algebra/ContinuousMonoidHom.lean

```diff
@@ -34,57 +34,59 @@ variable (F A B C D E : Type _) [Monoid A] [Monoid B] [Monoid C] [Monoid D] [Com
 
 /-- The type of continuous additive monoid homomorphisms from `A` to `B`.
 
-When possible, instead of parametrizing results over `(f : continuous_add_monoid_hom A B)`,
-you should parametrize over `(F : Type*) [continuous_add_monoid_hom_class F A B] (f : F)`.
+When possible, instead of parametrizing results over `(f : ContinuousAddMonoidHom A B)`,
+you should parametrize over `(F : Type*) [ContinuousAddMonoidHomClass F A B] (f : F)`.
 
-When you extend this structure, make sure to extend `continuous_add_monoid_hom_class`. -/
+When you extend this structure, make sure to extend `ContinuousAddMonoidHomClass`. -/
 structure ContinuousAddMonoidHom (A B : Type _) [AddMonoid A] [AddMonoid B] [TopologicalSpace A]
   [TopologicalSpace B] extends A →+ B where
-  continuous_toFun : Continuous to_fun
+  /-- Proof of continuity of the Hom. -/
+  continuous_toFun : @Continuous A B _ _ toFun
 #align continuous_add_monoid_hom ContinuousAddMonoidHom
 
 /-- The type of continuous monoid homomorphisms from `A` to `B`.
 
-When possible, instead of parametrizing results over `(f : continuous_monoid_hom A B)`,
-you should parametrize over `(F : Type*) [continuous_monoid_hom_class F A B] (f : F)`.
+When possible, instead of parametrizing results over `(f : ContinuousMonoidHom A B)`,
+you should parametrize over `(F : Type*) [ContinuousMonoidHomClass F A B] (f : F)`.
 
-When you extend this structure, make sure to extend `continuous_add_monoid_hom_class`. -/
-@[to_additive]
+When you extend this structure, make sure to extend `ContinuousAddMonoidHomClass`. -/
+@[to_additive "The type of continuous additive monoid homomorphisms from `A` to `B`."]
 structure ContinuousMonoidHom extends A →* B where
-  continuous_toFun : Continuous to_fun
+  /-- Proof of continuity of the Hom. -/
+  continuous_toFun : @Continuous A B _ _ toFun
 #align continuous_monoid_hom ContinuousMonoidHom
-#align continuous_add_monoid_hom ContinuousAddMonoidHom
 
 section
 
-/-- `continuous_add_monoid_hom_class F A B` states that `F` is a type of continuous additive monoid
+/-- `ContinuousAddMonoidHomClass F A B` states that `F` is a type of continuous additive monoid
 homomorphisms.
 
-You should also extend this typeclass when you extend `continuous_add_monoid_hom`. -/
-class ContinuousAddMonoidHomClass (A B : Type _) [AddMonoid A] [AddMonoid B] [TopologicalSpace A]
-  [TopologicalSpace B] extends AddMonoidHomClass F A B where
+You should also extend this typeclass when you extend `ContinuousAddMonoidHom`. -/
+-- porting note : Changed A B to outParam to help synthesizing order
+class ContinuousAddMonoidHomClass (A B : outParam (Type _)) [AddMonoid A] [AddMonoid B]
```

Comment:
Hmm, I'm a bit surprised that this didn't generate a whole new class of errors. That might be due to the fact that we haven't really ported files yet that use this class.
In general, I don't see how type class inference can guess what `A` and `B` are, if it doesn't know them already. So I'm pretty skeptical that they should be out params.
Was the synthesizing error discussed on zulip?

Reply:
I've modeled this after other classes of this type (like [ContinuousOrderHomClass](https://github.com/leanprover-community/mathlib4/blob/e28e13ff50994c498fbefcf048f887e622ac8f80/Mathlib/Topology/Order/Hom/Basic.lean#L53) and [AddMonoidHomClass](https://github.com/leanprover-community/mathlib4/blob/e28e13ff50994c498fbefcf048f887e622ac8f80/Mathlib/Algebra/Hom/Group.lean#L180) )

There's shouldn't be any files using this (at least not according to the portig page). I think the reason this works is because the typeclass hierarchy ends up at `FunLike F α fun _ => β`.

Does the above make sense or should I still start a thread in zulip?

Reply:
Sorry, you are completely right. I got confused. Of course the typeclass algorithm has the argument `F` to key on, and it can use that to determine `A` and `B`.
So all is fine!